### PR TITLE
Add Inline Comments for the C# sample

### PR DESCRIPTION
**Pull request by AhmedAbdoOrtiga**
#### Description of changes
- Add Inline Comments for the CSharpSeleniumWebdriverSample project.

#### Pull request checklist

- [ ] If this PR addresses an existing issue, it is linked: Fixes #0000
- [ ] `yarn test` passes in all affected samples
- [ ] Added any applicable tests

### CODE REVIEW

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 1]
```diff
@@ -6,57 +6,72 @@
...
         public void RunScanOnGivenElement(BrowserType browser)
         {
+            //  Intilaize the driver based on the browser type
             this.InitDriver(browser);
+            //  Load the test page
             LoadTestPage();
 
+            //  Analyze the unorder list element in the page, and get the results
             AxeResult results = RunScanOnGivenElementBySelector("ul");
+            //  Assert that the number of vilaztions that were found is that what you expect
```
Comment (devaradhanm) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: typo

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 2]
```diff
@@ -6,57 +6,72 @@
 using System;
 using System.IO;
 using Selenium.Axe;
-using OpenQA.Selenium.Chrome;
 
 namespace CSharpSeleniumWebdriverSample.Test
 {
     [TestClass]
-    [DeploymentItem("SamplePage.html")]
+    //  Install chromedriver to enable Chrome broswer and copy it to the same directory as the deployed test assemblies.
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: I don't think this is accurate; this line isn't causing chromedriver to be "installed". I think it would be more valuable for this comment to be describing why this attribute would be necessary even when the `Selenium.Webdriver.ChromeDriver` already causes the binary to be placed in the built directory without the attribute.

Honestly, I'd be okay with just removing all 3 of the DeploymentItem attributes here; I think I'm comfortable leaving the use cases they enable as out of scope for this sample.

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 3]
```diff
@@ -6,57 +6,72 @@
...
         [TestCleanup]
         public virtual void TearDown()
         {
+            //  You should quit the driver on TearDown which close every associated window.
             _webDriver?.Quit();
         }
 
 
         [TestMethod]
+        // Test on Chrome
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: I don't think this comment is adding any value; it's pretty apparent that the code is intended to enable testing on Chrome. I think a more valuable comment here might be a 1-sentence summary of how the `DataRow` attribute works with a link to MSTest docs explaining it in more detail, or maybe suggestions for ways besides `DataRow` to achieve the "write test once, run multiple times on different browsers" behavior

File: csharp-selenium-webdriver-sample/WebDriverFactory.cs [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 namespace CSharpSeleniumWebdriverSample
 {
+    // Enum for the browser types
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/WebDriverFactory.cs]: Comment isn't adding value

File: csharp-selenium-webdriver-sample/WebDriverFactory.cs [hunk 5]
```diff
@@ -21,13 +22,15 @@ public static IWebDriver GetWebDriver(BrowserType browser)
             switch (browser)
             {
                 case BrowserType.Chrome:
-                    // Check if Chrome web driver environment variable already defined so that Chrome version and the webdriver are the same, otherwise use Selenium.WebDriver.ChromeDriver to install it                    
+                    //  Check if Chrome web driver environment variable already defined so that Chrome version and the webdriver are the same, otherwise use Selenium.WebDriver.ChromeDriver to install it
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/WebDriverFactory.cs]: This comment is one of the better ones in that it's describing some of *why* the line of code exists, rather than just repeating what it's doing. I think it would be nice to expand this a little (how exactly does the environment variable ensure that the chrome version is in sync? what sets the environment variable? in what way does saying "CurrentDirectory" cause Selenium.WebDriver.ChromeDriver to install it?). You should be able to find a very similar line of code in the typescript sample that would be a good starting point

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 6]
```diff
@@ -6,57 +6,72 @@
...
             var selectedElement = _webDriver.FindElement(By.TagName(elementSelector));
 
+            //  Instruct the driver to analyze the current element, and return the result as AxeResult object
             return _webDriver.Analyze(selectedElement);
         }
 
 
         private void LoadTestPage()
         {
+            //  For simplicity, we're pointing our test browser directly to a static html file on disk.
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: It would be nice for this to describe in a little more depth how this would differ from a "real" project. There's a comment in a similar spot in the typescript sample to use as a starting point

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 7]
```diff
@@ -6,57 +6,72 @@
...
     [DeploymentItem("geckodriver.exe")]
     public class SampleTest
     {
+        //  Intance to control the browser.
         private IWebDriver _webDriver;
 
         [TestCleanup]
         public virtual void TearDown()
         {
+            //  You should quit the driver on TearDown which close every associated window.
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: Grammar ("which close")

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 8]
```diff
@@ -6,57 +6,72 @@
...
 {
     [TestClass]
-    [DeploymentItem("SamplePage.html")]
+    //  Install chromedriver to enable Chrome broswer and copy it to the same directory as the deployed test assemblies.
     [DeploymentItem("chromedriver.exe")]
+    //  Install geckodriver to enable Firefox brower and copy it to the same directory as the deployed test assemblies.
     [DeploymentItem("geckodriver.exe")]
     public class SampleTest
     {
+        //  Intance to control the browser.
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: Spelling on "Intance", but also I think it would be nice for this to explain a little more what the significance of this object is

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 9]
```diff
@@ -6,57 +6,72 @@
...
 
 
         [TestMethod]
+        // Test on Chrome
         [DataRow(BrowserType.Chrome)]
+        //  Test on Firefox
         [DataRow(BrowserType.Firefox)]
         public void RunScanOnGivenElement(BrowserType browser)
         {
+            //  Intilaize the driver based on the browser type
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: I don't think any of the comments from here through line 53 are adding much value; they're all just repeating the code directly underneath them. I'd definitely recommend looking over index.test.ts in the typescript sample for some examples of the types of comments I think would be more useful, describing either *why* particular bits of code are present or describing what's happening at a higher level (eg, whole tests)

File: csharp-selenium-webdriver-sample/CSharpSeleniumWebdriverSample.csproj [hunk 10]
```diff
@@ -1,10 +1,9 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/CSharpSeleniumWebdriverSample.csproj]: It would be nice to see comments added to `azure-pipelines.yml` as well, similar to the other sample (but if you'd like to keep that to a separate PR, that's fine)

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 11]
```diff
@@ -6,57 +6,72 @@
...
             string integrationTestTargetFile = Path.GetFullPath(@"SamplePage.html");
             string integrationTestTargetUrl = new Uri(integrationTestTargetFile).AbsoluteUri;
 
+            //  Instructs the driver to navigate the browser to the target URL, this should start the brower and open the page in a new window tab
             _webDriver.Navigate().GoToUrl(integrationTestTargetUrl);
         }
 
         private void InitDriver(BrowserType browser)
         {
+            //  Create a web driver based on the bowser type which is ehither Chrome of Firefox
```
Comment (DaveTryon) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: >   [](start = 14, length = 2)

It looks like the "2 spaces after the slashes" is an intentional pattern. I'm not sure if that's part of the style in the sample-pipelines repo, but it's not a pattern that we follow in AIWin...

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 12]
```diff
@@ -6,57 +6,76 @@
...
             this.InitDriver(browser);
             LoadTestPage();
 
+            // Analyze the a known element in the page, and get the results
             AxeResult results = RunScanOnGivenElementBySelector("ul");
+            // Assert that the number of violations that were found is that what you expect
             Assert.AreEqual(3, results.Violations.Length);
         }
 
         private AxeResult RunScanOnGivenElementBySelector(string elementSelector)
```
Comment (devaradhanm) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: method name says RunelementBySelector, but you are actually expecting tagname. may be use By.CssSelector

File: csharp-selenium-webdriver-sample/azure-pipelines.yml [hunk 13]
```diff
@@ -1,22 +1,55 @@
...
+# If you *don't* already have an Azure Pipelines build, we recommend following Azure Pipelines'
+# documentation at https://docs.microsoft.com/en-us/azure/devops/pipelines/create-first-pipeline
+# (using this sample as your starter code).
+#
+# For more information, see:
+# * https://docs.microsoft.com/en-us/azure/devops/pipelines/get-started
+# * https://docs.microsoft.com/en-us/azure/devops/pipelines/customize-pipeline
+
 pool:
+  # The 'windows-2019' and 'vs20xx-windows' vmImages come pre-installed with selenium drivers for common browsers.
```
Comment (devaradhanm) [inline csharp-selenium-webdriver-sample/azure-pipelines.yml]: why would selenium drivers be installed on vms? shouldnt this be part of the test libs? I also see this for the typescript project.

Comment (AhmedAbdoOrtiga) [reply csharp-selenium-webdriver-sample/azure-pipelines.yml]: If you look at the 2 links in the comments
 - https://docs.microsoft.com/en-us/azure/devops/pipelines/test/continuous-test-selenium
 - https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/win/Vs2019-Server2019-Readme.md
It's mentioned in them that Selenium is pre-installed on any windows image
why is it pre-installed? 
- I think so it'd be compatible with the browser versions installed on the same agent.
- The paths to the folders containing these drivers can be obtained from the environment variables named IEWebDriver (Internet Explorer), ChromeWebDriver (Google Chrome), and GeckoWebDriver (Firefox).
- That's why we do something like: var chromeDriverDirectory = Environment.GetEnvironmentVariable("ChromeWebDriver") ?? Environment.CurrentDirectory;
- So it should works fine on windows agents, other agents and local ,

Comment (dbjorge) [reply csharp-selenium-webdriver-sample/azure-pipelines.yml]: I agree that ideally we'd pin versions of all the test dependencies. The issue is that:

* there isn't a particularly simple/official way to pull down and install pinned versions of the actual browsers in a cross-platform manner; in particular, Google doesn't seem to offer officially supported channels for downloading non-current versions of Chrome.
* The webdrivers (particularly ChromeDriver) very explicitly only support being used with exactly-matching browser versions

The result is that the webdriver and the browser have to be kept in sync. When we debated how to do this in the typescript sample, we felt the simplest method for this sample to use was to use the synced webdriver/browser pair already present on the agents; since this sample is intended to be primarily about integrating axe, rather than setting up Selenium, we traded off the more robust version pinning for simplicity.

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 14]
```diff
@@ -6,57 +6,76 @@
...
+            // Assert that the number of violations that were found is that what you expect
             Assert.AreEqual(3, results.Violations.Length);
         }
 
-        private AxeResult RunScanOnGivenElementBySelector(string elementSelector)
+        private AxeResult RunScanOnGivenElementByTagName(string tagName)
         {
-            var selectedElement = _webDriver.FindElement(By.TagName(elementSelector));
+            // Find the first element matching the element tag name parameter
+            // * It acts as a sanity check that our browser automation setup basically works
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: Are these bullet points misplaced? These would make more sense if it were happening as part of `LoadTestPage`, but here I think the `FindElement` is just necessary to get the argument to `_webDriver.Analyze`

Comment (AhmedAbdoOrtiga) [reply csharp-selenium-webdriver-sample/SampleTest.cs]: I just deleted it, it'd make more sense it we get the main element like we do in the typescript

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 15]
```diff
@@ -6,57 +6,74 @@
...
         {
+            // You should quit the driver on TearDown which will close every associated window.
             _webDriver?.Quit();
         }
 
 
         [TestMethod]
+        // First DataRow Attribute will create a new instance of the test method and pass Chrome as a paramter, so it'll test on ChromDriver
+        // Second DataRow Attribute will create a new instance of the test method and pass Firefox as a paramter, so it'll test on FirefoxDriver
+        // Instead of useing DataRow Attribute, you can create a helper fucntion that takes the the broswer type as a paremter
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: spelling: "useing" -> "using"

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 16]
```diff
@@ -6,57 +6,74 @@
...
         [TestCleanup]
         public virtual void TearDown()
         {
+            // You should quit the driver on TearDown which will close every associated window.
             _webDriver?.Quit();
         }
 
 
         [TestMethod]
+        // First DataRow Attribute will create a new instance of the test method and pass Chrome as a paramter, so it'll test on ChromDriver
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: spelling: "paramter" -> "parameter" (there are a few other instances of this to ctrl+F for)

File: csharp-selenium-webdriver-sample/SampleTest.cs [hunk 17]
```diff
@@ -6,57 +6,74 @@
...
+        // Second DataRow Attribute will create a new instance of the test method and pass Firefox as a paramter, so it'll test on FirefoxDriver
+        // Instead of useing DataRow Attribute, you can create a helper fucntion that takes the the broswer type as a paremter
+        // Then call it twice with 2 differrent test methods for each browser type
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: spelling: differrent -> different

File: csharp-selenium-webdriver-sample/SampleTest.cs (hunk 16 above)
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: spelling: ChromDriver -> ChromeDriver

File: csharp-selenium-webdriver-sample/SampleTest.cs (hunk 15 above)
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/SampleTest.cs]: spelling: fucntion -> function

Comment (dbjorge) [reply csharp-selenium-webdriver-sample/SampleTest.cs]: There a few other spelling errors I missed in my first review that the VS Code "Code Spell Checker" extension caught; definitely recommend installing that and using it to check over the rest of the comments :)

File: csharp-selenium-webdriver-sample/WebDriverFactory.cs [hunk 18]
```diff
@@ -21,13 +21,29 @@ public static IWebDriver GetWebDriver(BrowserType browser)
...
+                    // if it's not set, we use Selenium.WebDriver.ChromeDriver nuget to install it in the current directory
+                    // chromedriver.exe does not appear in Solution Explorer, but it is copied to bin folder from package folder when the build process.
                     var chromeDriverDirectory = Environment.GetEnvironmentVariable("ChromeWebDriver") ?? Environment.CurrentDirectory;
+
+                    // Initializes a new ChromeDriver using the specified path to the directory containing ChromeDriver.exe.
                     return new ChromeDriver(chromeDriverDirectory);
 
                 case BrowserType.Firefox:
-                    // Check if Gecko web driver environment variable already defined so that Firefox version and the webdriver are the same, otherwise use Selenium.WebDriver.GeckoDriver to install it                    
+                    // So now we need to specify the FirefoxDriver directory path
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/WebDriverFactory.cs]: Most of this is identical to the Chrome case above; I think it can probably just reference the above, rather than repeating the giant comment block

File: csharp-selenium-webdriver-sample/WebDriverFactory.cs [hunk 19]
```diff
@@ -21,13 +21,29 @@ public static IWebDriver GetWebDriver(BrowserType browser)
             switch (browser)
             {
                 case BrowserType.Chrome:
-                    // Check if Chrome web driver environment variable already defined so that Chrome version and the webdriver are the same, otherwise use Selenium.WebDriver.ChromeDriver to install it                    
+                    // So now we need to specify the ChromeDriver directory path
+                    // We check first if the enivronemnt variable was set, if so we use the path value
+                    // Usually if you test your project on an azure pipleine, the selenium Chrome driver is pre-installed
+                    // The paths to the folder containing it can be obtained from the environment variables named ChromeWebDriver (Google Chrome).
+                    // So  it'd be compatible with the Chrome browser versions installed on the same agent.
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/WebDriverFactory.cs]: extra space between "So  it'd"

File: csharp-selenium-webdriver-sample/WebDriverFactory.cs [hunk 20]
```diff
@@ -21,13 +21,29 @@ public static IWebDriver GetWebDriver(BrowserType browser)
             switch (browser)
             {
                 case BrowserType.Chrome:
-                    // Check if Chrome web driver environment variable already defined so that Chrome version and the webdriver are the same, otherwise use Selenium.WebDriver.ChromeDriver to install it                    
+                    // So now we need to specify the ChromeDriver directory path
```
Comment (dbjorge) [inline csharp-selenium-webdriver-sample/WebDriverFactory.cs]: Consider updating this to refer to the pipelines docs about the environment variable and to focus more on the reasoning for *why* we're using them. Perhaps something like this?

```cs
// The ChromeWebDriver environment variable comes pre-set in the Azure Pipelines VM Image we
// specify in azure-pipelines.yml. ChromeDriver requires that you use *matching* versions of Chrome and
// the Chrome WebDriver; in the build agent VMs, using this environment variable will make sure that we use
// the pre-installed version of ChromeDriver that matches the pre-installed version of Chrome.
//
// See https://docs.microsoft.com/en-us/azure/devops/pipelines/test/continuous-test-selenium
//
// Environment.CurrentDirectory is where the Selenium.Webdriver.ChromeDriver NuGet package will place the
// version of the Chrome WebDriver that it comes with. We fall back to this so that if a developer working on
// the project hasn't separately installed ChromeDriver, the test will still be able to run on their machine.
```
